### PR TITLE
S4 ORM for Postgres

### PR DESCRIPTION
**Pull request by pinebit**

### CODE REVIEW

File: core/store/migrate/migrations/0171_add_s4_functions_table.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    version INT NOT NULL,
+    expiration BIGINT NOT NULL,
+    confirmed BOOLEAN NOT NULL,
+    payload BYTEA NOT NULL,
+    signature BYTEA NOT NULL,
+    updated_at TIMESTAMPTZ NOT NULL
+);
+
+CREATE UNIQUE INDEX functions_address_slot_id_idx ON "s4".functions(address, slot_id);
+CREATE INDEX functions_version_idx ON "s4".functions(version);
```
Comment (bolekk) [inline core/store/migrate/migrations/0171_add_s4_functions_table.sql:18]: Does this make sense? Do you ever do global lookups by version?

File: core/services/s4/postgres_orm.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+var _ ORM = (*postgres)(nil)
+
+const fields = "address, slot_id, version, expiration, confirmed, payload, signature, updated_at"
+
+func NewPostgresORM(db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig, tableName string) ORM {
+	return &postgres{
+		q:         pg.NewQ(db, lggr, cfg),
+		tableName: fmt.Sprintf(`"s4".%s`, tableName),
```
Comment (bolekk) [inline core/services/s4/postgres_orm.go:28]: schema name to a const?

File: core/services/s4/postgres_orm.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/sqlx"
+)
+
+type postgres struct {
```
Comment (bolekk) [inline core/services/s4/postgres_orm.go:16]: maybe "postgresORM"? It looks like most modules in core call this object just "orm".

File: core/services/s4/postgres_orm.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	q         pg.Q
+	tableName string
+}
+
+var _ ORM = (*postgres)(nil)
+
+const fields = "address, slot_id, version, expiration, confirmed, payload, signature, updated_at"
+
+func NewPostgresORM(db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig, tableName string) ORM {
```
Comment (bolekk) [inline core/services/s4/postgres_orm.go:25]: Isn't the convention to put logger last?

File: core/services/s4/postgres_orm.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+func (p postgres) Update(row *Row, qopts ...pg.QOpt) error {
+	q := p.q.WithOpts(qopts...)
+
+	stmt := fmt.Sprintf(`INSERT INTO %s as t (%s)
+	                     VALUES ($1, $2, $3, $4, $5, $6, $7, NOW())
+					     ON CONFLICT (address, slot_id)
+						 DO UPDATE SET version = EXCLUDED.version,
+									   expiration = EXCLUDED.expiration,
```
Comment (bolekk) [inline core/services/s4/postgres_orm.go:53]: some tab/space issue here

**Review by bolekk — APPROVED** [review]

File: core/services/s4/postgres_orm.go (hunk 4 above)
Comment (pinebit) [reply core/services/s4/postgres_orm.go:25]: I cannot remember such a convention.. I just followed `pg.NewQ` params order.

File: core/store/migrate/migrations/0171_add_s4_functions_table.sql (hunk 1 above)
Comment (pinebit) [reply core/store/migrate/migrations/0171_add_s4_functions_table.sql]: I thought it is good for Update, but looks like you are right, this is not needed for that context.

**Review by bolekk — DISMISSED** [review]

**Review by bolekk — APPROVED** [review]

**Pull request merged**